velox/functions/sparksql/Mask.cpp

```diff
@@ -0,0 +1,266 @@
+#include "velox/expression/VectorWriters.h"
+#include "velox/functions/lib/Re2Functions.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+/**
+ * Masks the characters of the given string value with the provided specific
+ * characters respectively.
+ *
+ * mask(string) -> string
+ * replaces upper-case characters with 'X', lower-case characters with 'x',
+ * and numbers with 'n'.
+ *
+ * mask(string, upperChar) -> string
+ * replaces upper-case characters with provided `upperChar`, lower-case
+ * characters with 'x', and numbers with 'n'.
+ * Upper case characters remain unmasked if provided `upperChar` is NULL.
+ *
+ * mask(string, upperChar, lowerChar) -> string
+ * replaces upper-case characters with provided `upperChar`, lower-case
+ * characters with provided `lowerChar` and numbers with 'n'.
+ * Upper case characters or lower-case characters would remain unmasked if
+ * provided `upperChar` or `lowerChar` is NULL.
+ *
+ * mask(string, upperChar, lowerChar, digitChar) -> string
+ * replaces upper-case characters with provided `upperChar`, lower-case
+ * characters with provided `lowerChar` and numbers with provided `digitChar`.
+ * Upper case characters, lower-case characters and numbers would remain
+ * unmasked if provided `upperChar`, `lowerChar` or `digitChar` is NULL.
+ *
+ * mask(string, upperChar, lowerChar, digitChar, otherChar) -> string
+ * replaces upper-case characters with provided `upperChar`, lower-case
+ * characters with provided `lowerChar` and numbers with provided `digitChar`
+ * and all other characters with provided `otherChar`.
+ * Upper case characters remain unmasekd if provided `upperChar` is NULL.
+ * Lower case characters remain unmasekd if provided `lowerChar` is NULL.
+ * Numbers remain unmasekd if provided `digitChar` is NULL.
+ * All other characters remain unmasked if provided `otherChar` is NULL.
+ */
+class MaskFunction final : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& /* outputType */,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    // Get the decoded vectors out of arguments.
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    DecodedVector* strings = decodedArgs.at(0);
+    DecodedVector* upperChars = args.size() >= 2 ? decodedArgs.at(1) : nullptr;
+    DecodedVector* lowerChars = args.size() >= 3 ? decodedArgs.at(2) : nullptr;
+    DecodedVector* digitChars = args.size() >= 4 ? decodedArgs.at(3) : nullptr;
+    DecodedVector* otherChars = args.size() >= 5 ? decodedArgs.at(4) : nullptr;
+    BaseVector::ensureWritable(rows, VARCHAR(), context.pool(), result);
+    auto* results = result->as<FlatVector<StringView>>();
+    auto getMaskedChar =
+        [&](DecodedVector* inputChars,
+            const VectorPtr& arg,
+            int index,
+            std::string_view maskedChar) -> std::optional<StringView> {
```

Comment:
std::string_view -> char
Let this function return std::optional<char>, and check for the input size here.
> VELOX_USER_CHECK(
            charOpt.value().size() == 1,
            std::string("Length of ") + charType + " should be 1");

Perhaps we don't need charType in the error message because their size should all be 1.